rust/src/utils/kmeans.rs

```diff
@@ -0,0 +1,306 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use arrow_arith::arithmetic::{add, divide_scalar};
+use arrow_array::{
+    cast::as_primitive_array, new_empty_array, types::Float32Type, Array, FixedSizeListArray,
+    Float32Array,
+};
+use arrow_schema::DataType;
+use arrow_select::concat::concat;
+use futures::stream::{self, repeat_with, StreamExt, TryStreamExt};
+use rand::prelude::*;
+use rand::{distributions::WeightedIndex, Rng, RngCore};
+
+use super::distance::l2_distance;
+use crate::Result;
+use crate::{arrow::*, Error};
+
+#[derive(Debug)]
+pub struct KMeansParams {
+    /// Max number of iterations.
+    max_iters: u32,
+
+    /// When the difference of mean distance to the centroids is less than this `tolerance`
+    /// threshold, stop the training.
+    tolerance: f32,
+
+    /// Run kmeans mulitple times and pick the best one.
+    redos: usize,
+}
+
+impl Default for KMeansParams {
+    fn default() -> Self {
+        Self {
+            max_iters: 50,
+            tolerance: 1e-4,
+            redos: 1,
+        }
+    }
+}
+
+/// KMeans implementation for Apache Arrow Arrays.
+#[derive(Debug, Clone)]
+pub struct KMeans {
+    /// Centroids for each of the k clusters.
+    ///
+    /// k * dimension.
+    pub centroids: Arc<FixedSizeListArray>,
+
+    /// The number of clusters
+    pub k: u32,
+}
+
+/// Initialize using kmean++, and returns the centroids of k clusters.
+async fn kmean_plusplus(
+    data: Arc<FixedSizeListArray>,
+    k: u32, /* dist_fn, rand_seed */
+    rng: &mut dyn RngCore,
+) -> KMeans {
+    assert!(data.len() > k as usize);
+    let dimension = data.value_length();
+
+    let mut kmeans = KMeans::empty(k, dimension);
+
+    let first = rng.gen_range(0..data.len());
+    let vector = data.value(first);
+    kmeans.centroids = FixedSizeListArray::try_new(vector, dimension)
+        .unwrap()
+        .into();
+
+    let mut seen = HashSet::new();
+    seen.insert(first);
+    for _ in 1..k {
+        let membership = kmeans.compute_membership(data.clone()).await;
+        let weights = WeightedIndex::new(&membership.distances).unwrap();
+        let mut chosen;
+        loop {
+            chosen = weights.sample(rng);
+            if !seen.contains(&chosen) {
+                seen.insert(chosen);
+                break;
+            }
+        }
+        let vector = data.value(chosen);
+        let new_vector: &Float32Array = as_primitive_array(&vector);
+
+        let centroids_array = kmeans.centroids.values();
```

Comment:
Maybe i should just keep `kmeans.centroids` as `Float32Array`, using `FixedSizeListArray` losing the type info, for example, `centroids_array` here is `Arc<dyn Array>` that's where it needs to cast back to `Float32Array` in the next line.

What about addressing these in the next PR that integrates with `create indexing`? It will make us see what way is more ergonomic.